python/cudf_polars/cudf_polars/experimental/expressions.py

```diff
@@ -0,0 +1,485 @@
+        for agg, column_chunk_list in zip(combine_aggs, column_chunk_lists, strict=True)
+    ]
+
+    if finalize:
+        # Perform optional BinOp on combined columns
+        dt, op_name = finalize
+        op = getattr(plc.binaryop.BinaryOperator, op_name)
+        col = Column(plc.binaryop.binary_operation(*(c.obj for c in combined), op, dt))
+    else:
+        assert len(combined) == 1
+        col = combined[0]
+
+    return col.rename(name)
+
+
+def shuffle_child(
+    child: IR,
+    on: Expr,
+    named_expr: NamedExpr,
+    child_partition_info: PartitionInfo,
+    config_options: ConfigOptions,
+) -> tuple[IR, MutableMapping[Any, Any]]:
+    """
+    Shuffle the child IR on a single expression.
+
+    Parameters
+    ----------
+    child
+        Child IR we are shuffling.
+    on
+        Single Expr to shuffle on.
+    named_expr
+        Outer NamedExpr needing this shuffle.
+    child_partition_info
+        IR partitioning information.
+    config_options
+        GPU engine configuration options.
+
+    Returns
+    -------
+    shuffled
+        New (shuffled) child. The corresponding DataFrame
+        container will only contain the columns associated
+        with the shuffle keys (``on``).
+    graph
+        Task graph needed to shuffle child.
+    """
+    graph: MutableMapping[Any, Any] = {}
+
+    if child_partition_info.partitioned_on == (
+        named_expr.reconstruct(on),
+    ):  # pragma: no cover
+        # No shuffle necessary
+        return child, graph
+
+    child_partition_info = PartitionInfo(count=child_partition_info.count)
+    shuffle_on = (
+        named_expr.reconstruct(
+            FusedExpr(
+                on.dtype,
+                on,
+                *[c for c in named_expr.value.children if isinstance(c, FusedExpr)],
+            )
+        ),
+    )
+
+    pi: MutableMapping[IR, PartitionInfo] = {child: child_partition_info}
+    schema = {col.name: col.dtype for col in traversal([on]) if isinstance(col, Col)}
+    if set(schema) != set(child.schema):
+        # Drop unnecessary columns before the shuffle
+        child = Select(
+            schema,
+            shuffle_on,
+            False,  # noqa: FBT003
+            child,
+        )
+        pi[child] = child_partition_info
+        graph.update(generate_ir_tasks(child, pi))
+    shuffled = Shuffle(
+        schema,
+        shuffle_on,
+        config_options,
+        child,
+    )
+    pi[shuffled] = child_partition_info
+    graph.update(generate_ir_tasks(shuffled, pi))
+
+    return shuffled, graph
+
+
+def make_agg_graph(
+    named_expr: NamedExpr,
+    expr_partition_counts: MutableMapping[Expr, int],
+    child: IR,
+    child_partition_info: PartitionInfo,
+) -> MutableMapping[Any, Any]:
+    """Build a FusedExpr aggregation graph."""
+    expr = named_expr.value
+    assert isinstance(expr, FusedExpr)
+    agg = expr.sub_expr
+    assert isinstance(agg, Agg), f"Expected Agg, got {agg}"
+    _check_sub_expr(agg)
```

Comment:
It seems like we're checking these aggregations multiple times, since when we decompose to produce fusedexprs we also call this check.